gateway/gateway.go

```diff
@@ -0,0 +1,366 @@
+
+	headers := []string{"Miner", "Price", "Min Piece Size", "Timestamp", "Expiry"}
+
+	rows := make([][]interface{}, len(index.Storage))
+	i := 0
+	for _, ask := range index.Storage {
+		rows[i] = []interface{}{
+			ask.Miner,
+			ask.Price,
+			ask.MinPieceSize,
+			timeToString(uint64ToTime(ask.Timestamp)),
+			timeToString(uint64ToTime(ask.Expiry)),
+		}
+		i++
+	}
+
+	c.HTML(http.StatusOK, "/public/html/asks.gohtml", gin.H{
+		"MenuItems": menuItems,
+		"Title":     "Available Asks",
+		"Subtitle":  subtitle,
+		"Headers":   headers,
+		"Rows":      rows,
+	})
+}
+
+func (g *Gateway) minersHandler(c *gin.Context) {
+	menuItems := makeMenuItems(1)
+
+	index := g.minerIndex.Get()
+
+	metaSubtitle := fmt.Sprintf("%v miners online, %v miners offline", index.Meta.Online, index.Meta.Offline)
+	metaHeaders := []string{"Miner", "Location", "Online", "User Agent", "Updated"}
+	metaRows := make([][]interface{}, len(index.Meta.Info))
+	i := 0
+	for id, meta := range index.Meta.Info {
+		metaRows[i] = []interface{}{
+			id,
+			meta.Location.Country,
+			meta.Online,
+			meta.UserAgent,
+			timeToString(meta.LastUpdated),
+		}
+		i++
+	}
+
+	chainSubtitle := fmt.Sprintf("Last updated %v", timeToString(uint64ToTime(index.Chain.LastUpdated)))
+	chainHeaders := []string{"Miner", "Power", "Relative"}
+	chainRows := make([][]interface{}, len(index.Chain.Power))
+	i = 0
+	for id, power := range index.Chain.Power {
+		chainRows[i] = []interface{}{
+			id,
+			power.Power,
+			power.Relative,
+		}
+		i++
+	}
+
+	c.HTML(http.StatusOK, "/public/html/miners.gohtml", gin.H{
+		"MenuItems": menuItems,
+		"MetaData": gin.H{
+			"Title":    "Miner Metadata",
+			"Subtitle": metaSubtitle,
+			"Headers":  metaHeaders,
+			"Rows":     metaRows,
+		},
+		"ChainData": gin.H{
+			"Title":    "Miner On-Chain Data",
+			"Subtitle": chainSubtitle,
+			"Headers":  chainHeaders,
+			"Rows":     chainRows,
+		},
+	})
+}
+
+func (g *Gateway) slashingHandler(c *gin.Context) {
+	menuItems := makeMenuItems(2)
+
+	index := g.slashingIndex.Get()
+
+	subtitle := fmt.Sprintf("Current tip set key: %v", index.TipSetKey)
+
+	headers := []string{"Miner", "Slashed Epochs"}
+
+	rows := make([][]interface{}, len(index.Miners))
+	i := 0
+	for id, slashes := range index.Miners {
+		epochs := make([]string, len(slashes.Epochs))
+		for j, epoch := range slashes.Epochs {
+			epochs[j] = string(epoch)
+		}
+		rows[i] = []interface{}{
+			id,
+			strings.Join(epochs, ", "),
+		}
+		i++
+	}
+
+	c.HTML(http.StatusOK, "/public/html/slashing.gohtml", gin.H{
+		"MenuItems": menuItems,
+		"Title":     "Miner Slashes",
+		"Subtitle":  subtitle,
+		"Headers":   headers,
+		"Rows":      rows,
+	})
+}
+
+func (g *Gateway) reputationHandler(c *gin.Context) {
+	menuItems := makeMenuItems(3)
+
+	topMiners, err := g.reputationModule.GetTopMiners(30)
+	if err != nil {
+		g.renderError(c, http.StatusInternalServerError, err)
```

Comment:
Nit: maybe make `30` a const, and use it on line 282 to make the text future-proof.

Reply:
cool, will do